src/stdlib_sparse_conversion.fypp

```diff
@@ -0,0 +1,452 @@
+        
+        do i = 1, CSR%nrows
+            do j = CSR%rowptr(i), CSR%rowptr(i+1)-1
+                COO%index(1:2,j) = [i,CSR%col(j)]
+            end do
+        end do
+    end subroutine
+
+    #:endfor
+
+    #:for k1, t1, s1 in (KINDS_TYPES)
+    subroutine csr2sellc_${s1}$(CSR,SELLC,chunk)
+        !> csr2sellc: This function enables transfering data from a CSR matrix to a SELL-C matrix
+        !> This algorithm was gracefully provided by Ivan Privec and adapted by Jose Alves
+        type(CSR_${s1}$), intent(in)      :: CSR
+        type(SELLC_${s1}$), intent(inout) :: SELLC
+        integer, intent(in), optional :: chunk
+        ${t1}$, parameter :: zero = zero_${s1}$
+        integer :: i, j, num_chunks
+
+        if(present(chunk)) SELLC%chunk_size = chunk
+
+        SELLC%nrows = CSR%nrows; SELLC%ncols = CSR%ncols
+        SELLC%base  = CSR%base;  SELLC%storage = CSR%storage
+        associate( nrows=>SELLC%nrows, ncols=>SELLC%ncols, nnz=>SELLC%nnz, &
+        &         chunk_size=>SELLC%chunk_size     )
+        !-------------------------------------------
+        ! csr rowptr to SELL-C chunked rowptr
+        num_chunks = (nrows + chunk_size - 1)/chunk_size
+        allocate( SELLC%rowptr(num_chunks+1) )
+        block
+            integer :: cidx, rownnz, chunknnz
+            SELLC%rowptr(1) = 1
+            cidx = 1
+            do i = 1, nrows, chunk_size
+                chunknnz = 0
+                ! Iterate over rows in a given chunk
+                do j = i, min(i+chunk_size-1,nrows)
+                    rownnz = CSR%rowptr(j+1) - CSR%rowptr(j)
+                    chunknnz = max(chunknnz,rownnz)
+                end do
+                SELLC%rowptr(cidx+1) = SELLC%rowptr(cidx) + chunknnz
+                cidx = cidx + 1
+            end do
+            nnz = SELLC%rowptr(num_chunks+1) - 1
+        end block
+        !-------------------------------------------
+        ! copy values and colum index
+        allocate(SELLC%col(chunk_size,nnz), source = -1)
+        allocate(SELLC%data(chunk_size,nnz), source = zero )
+        block
+            integer :: lb, ri, iaa, iab, rownnz
+            do i = 1, num_chunks
+
+                lb = SELLC%rowptr(i)
+
+                ! Loop over rows of a chunk
+                do j = (i-1)*chunk_size + 1, min(i*chunk_size,nrows)
+    
+                    ri = j - (i - 1)*chunk_size
+                    
+                    rownnz = CSR%rowptr(j+1) - CSR%rowptr(j) - 1
+                    iaa    = CSR%rowptr(j)
+                    iab    = CSR%rowptr(j+1) - 1
+                    
+                    SELLC%col(ri,lb:lb+rownnz)  = CSR%col(iaa:iab)
+                    SELLC%data(ri,lb:lb+rownnz) = CSR%data(iaa:iab)
+                
+                end do
+            end do
+         end block
+        end associate
+    end subroutine
+
+    #:endfor
+
+    recursive subroutine quicksort_i(a, first, last)
```

Comment:
Could the routines from `stdlib_sorting` be used instead?

Reply:
For the case of sorting only the index array that was easy. I'm still thinking how to do it for the case when wanting to make the data follow the sorting of the index without creating temporal arrays.

Reply:
Salvatore Filippone has written elsewhere that quicksort doesn't tend to be the most suitable for sorting indexes. (This can of course vary from matrix to matrix.) Not to block this nice PR, just saying that in the future this can be revisited for better performance. stdlib is a reference implementation first and foremost.

Reply:
> how to do it for the case when wanting to make the data follow the sorting of the index without creating temporal arrays.

To sort a list, and swap data according to the sort, this is the way I do it, you're welcome to use this template Jose:

```Fortran
    ! Quicksort, and swap list accordingly
    #:for lt,ln in LIST_TYPES_NAMES
    pure recursive subroutine ${name}$_quicksort_andlist_${ln}$(array,list,down)

      ${type}$, dimension(:), intent(inout) :: array
      ${lt}$, dimension(size(array)), intent(inout) :: list
      logical, optional, intent(in) :: down

      integer(ISIZE), parameter :: SMALL_SIZE = 20
      integer(ISIZE) :: i, j, n
      ${type}$  :: chosen
      logical :: descending

      descending = .false.; if (present(down)) descending = down
      n = size(array,kind=ISIZE)

      choose_sorting_algorithm: if (n>1 .and. n <= SMALL_SIZE) then

         ! Use interchange sort for small lists
         do i = 1, n - 1
            do j = i + 1, n
               if (toBeSwapped(array(i),array(j),.false.)) then
                   call swap_data_${name}$(array(i),array(j))
                   call swap_list(list(i),list(j))
               end if
            end do
         end do

      elseif (n>SMALL_SIZE) then
         ! Use partition (quick) sort if the array is big
         chosen = array(int(n/2))
         i = 0
         j = n + 1

         scan_lists: do
            ! Scan array from left end
            ! until element >= chosen is found
            scan_from_left: do
              i = i + 1
              if (toBeSwapped(array(i),chosen,.true.) .or. i==n) exit scan_from_left
            end do scan_from_left

            ! Scan array from right end
            ! until element <= chosen is found

            scan_from_right: do
               j = j - 1
               if (toBeSwapped(chosen,array(j),.true.) .or. j==1) exit scan_from_right
            end do scan_from_right

            swap_element: if (i < j) then
                ! Swap two out of place elements
                call swap_data_${name}$(array(i),array(j))
                call swap_list(list(i),list(j))
            else if (i == j) then
                i = i + 1
                exit
            else
                exit
            endif swap_element

         end do scan_lists

         if (1 < j) call ${name}$_quicksort_andlist_${ln}$(array(:j),list(:j),down)
         if (i < n) call ${name}$_quicksort_andlist_${ln}$(array(i:),list(i:),down)

      end if choose_sorting_algorithm ! test for small array

      contains
      
         elemental subroutine swap_list(a,b)
            ${lt}$, intent(inout) :: a,b
            ${lt}$ :: tmp
            tmp = a
            a = b
            b = tmp
         end subroutine swap_list

         elemental subroutine swap_data_${name}$(a,b)
            ${type}$, intent(inout) :: a, b
            ${type}$                :: tmp
            tmp = a
            a   = b
            b   = tmp
         end subroutine swap_data_${name}$

         elemental logical function toBeSwapped(a,b,orEqual)
            ${type}$, intent(in) :: a,b
            logical, intent(in) :: orEqual

            toBeSwapped = merge(is_before(a,b),is_after(a,b),descending)
            if (orEqual .and. a==b) toBeSwapped = .true.

         end function toBeSwapped

    end subroutine ${name}$_quicksort_andlist_${ln}$
    
    #:endfor
```

In this template: 
- `${type}$` is the array type (in your case, the integer column indices) 
- `${name}$` its template name
- `${lt}$` is the list type (in your case, `real(sp)` if you have real matrix data, etc.)
- we're working with numbers, so you can replace templates `is_before` with simpler comparisons `a<b` etc

Reply:
@ivan-pi Yes, that's why I used a combined sort: count_sort for the first level, then quicksort for the smaller chuncks for each row. Is the most efficient approach I have found so far. 
Both: in the sorting interfaces from stdlib I see that for making a secondary data be reordered following the sorting of something else, it is needed to create a temporal index mapping https://stdlib.fortran-lang.org/page/specs/stdlib_sorting.html "Sorting a related rank one array". This would be detrimental as is for large sparse matrices. That's why a included a quicksort that takes in the array to be sorted and an extra data buffer. @perazz the code you propose is very neat, the question is: should this be made available as a sorting interface elsewhere from stdlib and use it here as a particular application? (coming to your original comment of using stdlib facilities instead of locally cooked ones)

Reply:
@jalvesz I agree that no temporaries should be made. 
The way we dealt with this for algebra error control was to define a local error handler, in a way that can later be promoted to the whole stdlib. 

So one way to do this could be to write good code (e.g. fully templated for integer/real arguments of either array or list) here, so that it can nearly automatically be promoted to be an `stdlib_sorting` method later.